### PR TITLE
MULE-15427: Allow developer to declare an element as Required for Met…

### PR DESCRIPTION
**Pull request by estebanwasinger**
…adata

### CODE REVIEW

File: modules/extensions-spring-support/src/test/resources/models/petstore.json [hunk 1]
```diff
@@ -18357,9 +18369,13 @@
       "name": "PartialPetAdoptionSource",
       "description": "",
       "modelProperties": {
-        "org.mule.runtime.extension.api.property.TypeResolversInformationModelProperty": {
+        "typeResolversInformation": {
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/resources/models/petstore.json]: aren't you breaking BP with this?

Comment (estebanwasinger) [reply modules/extensions-spring-support/src/test/resources/models/petstore.json]: nop, this was added on 1.2.0 (current SNAPSHOT version) so we are still able to fix it

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/RequiredForMetadataDeclarationEnricher.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+      @Override
+      protected void onConnectionProvider(ConnectedDeclaration owner, ConnectionProviderDeclaration declaration) {
+        registerRequiredParametersForMetadata(declaration);
+      }
+    }.walk(extensionLoadingContext.getExtensionDeclarer().getDeclaration());
+  }
+
+  private <T extends BaseDeclaration & WithParametersDeclaration> void registerRequiredParametersForMetadata(T declaration) {
+    List<String> parametersRequiredForMetadata = getParametersRequiredForMetadata(declaration)
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/RequiredForMetadataDeclarationEnricher.java]: is weird that you create the stream and then keep modifying it outside to get the list. Why no just returning the names list?

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/enricher/RequiredForMetadataDeclarationEnricherTestCase.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    DefaultJavaModelLoaderDelegate loader = new DefaultJavaModelLoaderDelegate(MetadataExtension.class, getProductVersion());
+    ExtensionDeclarer declarer =
+        loader.declare(new DefaultExtensionLoadingContext(getClass().getClassLoader(), getDefault(emptySet())));
+    new RequiredForMetadataDeclarationEnricher()
+        .enrich(new DefaultExtensionLoadingContext(declarer, this.getClass().getClassLoader(), getDefault(emptySet())));
+    declaration = declarer.getDeclaration();
+  }
+
+  @Test
+  public void connectionProviderWithRequiredForMetadataParameterGetsEnriched() {
```
Comment (alegmarra) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/enricher/RequiredForMetadataDeclarationEnricherTestCase.java]: looks like you at least need two more tests: 
- config with required
- connection without required

Comment (alegmarra) [reply modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/enricher/RequiredForMetadataDeclarationEnricherTestCase.java]: (probably better doing unit+mock than using full extension)

Comment (estebanwasinger) [reply modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/enricher/RequiredForMetadataDeclarationEnricherTestCase.java]: 1) 👍 
2) is a mocked extension.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/DynamicMetadataDeclarationEnricher.java [hunk 4]
```diff
@@ -146,11 +147,15 @@ private void declareResolversInformation(ComponentDeclaration<? extends Componen
...
         String keysResolver = metadataScope.getKeysResolver().get().getResolverName();
+
+        //TODO MULE-15638 - Once Metadata API 2.0 is implemented we will know better if the resolver requires or not a connection of config.
         declaration.addModelProperty(new TypeResolversInformationModelProperty(categoryName,
                                                                                inputResolversByParam,
                                                                                outputResolver,
                                                                                attributesResolver,
-                                                                               keysResolver));
+                                                                               keysResolver,
+                                                                               declaration.isRequiresConnection(),
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/DynamicMetadataDeclarationEnricher.java]: this should come from the annotation

**Pull request merged**